keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,527 @@
+Each of those methods will accept log messages as a first argument and a list of key-values pairs as a variadic second argument.
+This approach allows incremental adoption of structured logging without converting ALL of kubernetes to a new API at one time.
+
+Declaration:
+
+```go
+package klog
+
+func InfoS(msg string, keysAndValues ...interface{})
+func ErrorS(err error, msg string, keysAndValues ...interface{})
+```
+
+Example:
+
+```go
+klog.InfoS("Pod status updated", "pod", "kubedns", "status", "ready")
+```
+
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kubedns" status="ready"
+```
+
+And
+
+```go
+klog.ErrorS(err, "Failed to update pod status")
+```
+
+That would result in log
+```
+E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+```
+
+### References to Kubernetes objects
+
+To address problem of inconsistency in referencing of Kubernetes objects, new klog methods will include additional logic for handling their formatting. Kubernetes objects will be recognized by checking for `ObjectMeta` structure.
+
+Namespaced objects will be represented by joining their name and namespace with `/` character. Pattern: `<namespace>/<name>`, example for pod: `kube-system/kubedns`
+
+Non-namespaced objects will be represented by their name. Pattern: `<name>`, example for node: `cluster1-vm-72x33b8p-34jz`
+
+Example:
+```go
+pod := corev1.Pod{Name: "kubedns", Namespace: "kube-system", ...}
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+```
+Selecting key under which put objects will be left to caller as defining log schema will need more discussion and is outside of scope of this KEP.
+
+### Selecting most important logs
+
+As migration to new message structure will be done manually we will be focusing on logs that have the greatest impact for log querying and processing. We will be focusing on log messages Proposed plan of measurements:
+
+* Using log data from single run of kubernetes [gce-master-scale-performance](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance) tests
+* Based on logs generated on master node
+* Aggregate from components:
+  * kube-controller-manager
+  * kube-scheduler
+  * kube-apiserver
+  * kubelet
+* Only for log lines generated by klog (e.g. prefix `I0129 03:26:19.857798       1 garbagecollector.go:517]`)
+* Identified by code line that generated log (e.g. `garbagecollector.go:517`)
+
+We assume that exact details of taking measurements can be improved, but overall methodology is solid. With those criteria we concluded that covering 99.9% of logs on master node are generated by 22 log messages. Exact list of those logs is provided in the detail design section.
```

Comment:
It seems like this is just telling you where the noise is, not where the value is.  Why assume that today's noise is where the value lies.

Reply:
The volume of logs produced by each klog call is just a proxy for the importance of those logs but indeed it does not mean that those logs are the most valuable.
If you can provide any other criteria to select logs which should be migrated first we might include it in this proposal.